mgr/dashboard: fix error while accessing roles tab when policy attached

afreen23 opened the pull request:
Fixes https://tracker.ceph.com/issues/64270

### Issue:
Accessing Object->Users-Roles tab causing 500 internal servor error. This is due to the "PermissionPolicies" which are attached to role and backend was not handling this field for rgw roles.
![image](https://github.com/ceph/ceph/assets/25664409/7ea93fe3-2b10-4586-803d-293999429137)


### Fix:
Added "PermissionPolicies" as the valid field in backend and updated frontend to render the attached policy in formatted JSON
![Screenshot from 2024-02-06 14-46-12](https://github.com/ceph/ceph/assets/25664409/1330e218-9591-4f12-9d3b-86bf51e6b567)


## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [x] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

afreen23 commented:
jenkins test make check

nizamial09 commented on src/pybind/mgr/dashboard/frontend/src/app/shared/datatable/crud-table/crud-table.component.html:
```diff
@@ -36,7 +36,7 @@
             <td i18n
                 class="bold">{{ column }}</td>
-            <td> {{ expandedRow[column] }} </td>
+            <td><pre>{{ customJSONStringify(expandedRow[column]) }}</pre></td>
```
it is generally not advised to use the function calls in templates because of the way angular [change detection is triggered](https://medium.com/showpad-engineering/why-you-should-never-use-function-calls-in-angular-template-expressions-e1a50f9c0496). 

Maybe we can move this logic or call this from `processMeta` function instead and do the processing if its required and update the `meta.detail_columns` with the newer values and then pass these updated value to the template?

Also this change means that we are passing `expandedRow[column]` through the json stringify even if it is not meant to be. Can we do this conditionally?

afreen23 replied:
> it is generally not advised to use the function calls in templates because of the way angular [change detection is triggered](https://medium.com/showpad-engineering/why-you-should-never-use-function-calls-in-angular-template-expressions-e1a50f9c0496). You can test this out by putting a console log in the function and see how many times its getting called
> 
> Maybe we can move this logic or call this from `processMeta` function instead and do the processing if its required and update the `meta.detail_columns` with the newer values and then pass these updated value to the template?
> 
I can check that surely.. makes sense if its not efficient.
I liked the idea of moving to `processMeta` and I can look into that...
> Also this change means that we are passing `expandedRow[column]` through the json stringify even if it is not meant to be. Can we do this conditionally?
I have changed the JSON.stringify replacer function, If you check the other file, if its a non-string, non-json, it will be returned as is.

The reason for doing this was first I tried `json` pipe which actually not taking care of escape sequences and was rendering them as it is. In a nutshell, it was non formatted json
I tried conditional way as well, and it was hardcoding certain fields or checking type of objects..
The `expandedRow[column]` is something coming as array of objects or strings, or objects.

afreen23 replied:
```
 /**
   * Checks strings for JSON value
   * If yes, returns the parsed value
   * If no, returns the original string
   *
   * For non-strings, return as it is
   **/
  customJSONStringify(rowValue: any) {
    return JSON.stringify(
      rowValue,
      (_key: string, value: any) => {
        if (typeof value === 'string') {
          try {
            const parsedValue = JSON.parse(value); 
            return parsedValue; 
          } catch (e) {
            return value; // returns string but non-json object as it is
          }
        }
        return value; // returns non-string values as it is
      },
      2
    );
  }
}
```
I have added a custom replacer function, so only valid values which are meant to be parsed will go through..

nizamial09 replied:
yup, i agree. probably a generic pipe could be useful later on for us.

afreen23 commented:
Hi @nizamial09 ,

I have updated the PR , combining all of your comments. PTAL and suggest if there is a better approach, I have come up with this current working now!

Summary of changes:
- `processMeta` is processing only meta data like column names, actions, etc. 
- we are processing live data asynchronously in html file 
- The only place where we have access to live data is in `this.expandedRow` . Therefore I moved data formatting inside that function.
- JSON.stringifywith the replacer function is recursively checking for objects and stringifying them so I kept it.
That is helpful for proper formatting on UI. I am also amazed at that since without replacer function its not going for objects properties and leaves them as it is [See below screenshot, PolicyValues are not formatted when used without a replacer function]
![Screenshot from 2024-02-07 21-14-26](https://github.com/ceph/ceph/assets/25664409/93f1ac49-7b9e-4eee-8291-c6b6542a7581)
I assume its the JSON.stringify algorithm, one of the many weird things with js..

Still, if there is any better approach for formatting deep nested json objects on UI let me know.
Will be glad to know!

Thanks!

afreen23 commented:
jenkins retest this please

nizamial09 reviewed (approved):
Thank you @afreen23. LGTM!

One thing that we need to do later on is to implement the ability to attach policies to a role from the UI. While testing it, I saw that we are missing that feature.

afreen23 commented:
> and btw, since you are fixing the bug here I think the title of the PR should clearly represent that. `mgr/dashboard: fix error while accessing roles tab when policy attached`.
> 
> and there are some unit-test failures.
> 
> ```
> [lint:html    ] 
> [lint:html    ] [htmllint] found 0 errors out of 212 files
> [lint:html    ] All files pass linting
> [lint:prettier] src/app/shared/datatable/crud-table/crud-table.component.ts
> [lint:eslint  ] 
> [lint:eslint  ] Linting "ceph-dashboard"...
> [lint:eslint  ] 
> [lint:eslint  ] All files pass linting.
> [lint:eslint  ] 
> 
> Try running 'npm run fix' to fix some linting errors. Some errors might need a manual fix.
> ```

Done with these...thanks!

afreen23 commented:
jenkins test dashboard cephadm
jenkins test dashboard
jenkins test make check arm64

afreen23 commented:
jenkins test make check arm64

afreen23 commented:
jenkins test dashboard

afreen23 commented:
jenkins test dashboard cephadm

afreen23 commented:
jenkins test dashboard

afreen23 commented:
jenkins test dashboard
